File: apps/api/tessera_api/api/templates.py
"""Маршруты шаблонов страниц."""

from __future__ import annotations

import uuid

import msgspec
from litestar import Controller, Request, post
from litestar.di import NamedDependency
from sqlalchemy.ext.asyncio import AsyncSession

from tessera_api.api.guards import Principal
from tessera_api.domain.errors import not_found
from tessera_api.infrastructure.models import User, Workspace
from tessera_api.services.templates import TemplateService


class TemplateIdRequest(msgspec.Struct):
    templateId: uuid.UUID  # noqa: N815 — имя поля из v1


class CreateTemplateRequest(msgspec.Struct):
    title: str
    description: str | None = None
    icon: str | None = None
    content: dict | None = None
    spaceId: uuid.UUID | None = None  # noqa: N815 — имя поля из v1


class UpdateTemplateRequest(msgspec.Struct):
    templateId: uuid.UUID  # noqa: N815 — имя поля из v1
    title: str | None = None
    description: str | None = None
    icon: str | None = None
    content: dict | None = None
    #: Перенос в другую область. Отдельное поле, потому что пустой `spaceId`
    #: означает «шаблон рабочего пространства», а не «область не меняем».
    moveToSpaceId: uuid.UUID | None = None  # noqa: N815
    move: bool = False


class UseTemplateRequest(msgspec.Struct):
    templateId: uuid.UUID  # noqa: N815 — имя поля из v1
    spaceId: uuid.UUID  # noqa: N815 — имя поля из v1
    parentPageId: uuid.UUID | None = None  # noqa: N815 — имя поля из v1


async def _actor(session: AsyncSession, principal: Principal) -> tuple[User, Workspace]:
    user = await session.get(User, principal.user_id)
    workspace = await session.get(Workspace, principal.workspace_id)
    if user is None or workspace is None:
        raise not_found("error.auth.account_unavailable")
    return user, workspace


class TemplateController(Controller):
    path = "/api/templates"

    @post("/")
    async def list_templates(
        self, request: Request, db_session: NamedDependency[AsyncSession]
    ) -> list[dict]:
        principal: Principal = request.scope["principal"]
        user, workspace = await _actor(db_session, principal)
        return await TemplateService(db_session).list(user, workspace.id)

    @post("/info")
    async def info(
        self,
        data: TemplateIdRequest,
        request: Request,
        db_session: NamedDependency[AsyncSession],
    ) -> dict:
        principal: Principal = request.scope["principal"]
        user, workspace = await _actor(db_session, principal)
        return await TemplateService(db_session).info(data.templateId, user, workspace)

    @post("/create")
    async def create(
        self,
        data: CreateTemplateRequest,
        request: Request,
        db_session: NamedDependency[AsyncSession],
    ) -> dict:
        principal: Principal = request.scope["principal"]
        user, workspace = await _actor(db_session, principal)
        return await TemplateService(db_session).create(
            user=user,
            workspace=workspace,
            title=data.title,
            description=data.description,
            icon=data.icon,
            content=data.content,
            space_id=data.spaceId,
        )

    @post("/update")
    async def update(
        self,
        data: UpdateTemplateRequest,
        request: Request,
        db_session: NamedDependency[AsyncSession],
    ) -> dict:
        principal: Principal = request.scope["principal"]
        user, workspace = await _actor(db_session, principal)
        return await TemplateService(db_session).update(
            template_id=data.templateId,
            user=user,
            workspace=workspace,
            title=data.title,
            description=data.description,
            icon=data.icon,
            content=data.content,
            space_id=data.moveToSpaceId,
            move=data.move,
        )

    @post("/delete")
    async def delete(
        self,
        data: TemplateIdRequest,
        request: Request,
        db_session: NamedDependency[AsyncSession],
    ) -> dict:
        principal: Principal = request.scope["principal"]
        user, workspace = await _actor(db_session, principal)
        await TemplateService(db_session).delete(data.templateId, user, workspace)
        return {"success": True}

    @post("/use")
    async def use(
        self,
        data: UseTemplateRequest,
        request: Request,
        db_session: NamedDependency[AsyncSession],
    ) -> dict:
        principal: Principal = request.scope["principal"]
        user, workspace = await _actor(db_session, principal)
        return await TemplateService(db_session).use(
            template_id=data.templateId,
            user=user,
            workspace=workspace,
            space_id=data.spaceId,
            parent_page_id=data.parentPageId,
        )

File: apps/web/src/lib/features/template/services/templates.ts
import { post } from '$lib/api/client';

export type Template = {
  id: string;
  title: string;
  description: string | null;
  icon: string | null;
  spaceId: string | null;
  creatorId: string | null;
  createdAt: string;
  updatedAt: string;
};

/** Шаблон вместе с содержимым. Список его не несёт: он открывается ради выбора. */
export type TemplateBody = Template & { content: unknown };

/** Созданная по шаблону страница. */
export type MadePage = {
  id: string;
  slugId: string;
  title: string | null;
  icon: string | null;
  spaceId: string;
};

/** Страница перечня и место, откуда продолжать. */
export type TemplatePage = {
  items: Template[];
  meta: { nextCursor: string | null };
};

/**
 * Шаблоны, доступные человеку.
 *
 * Отбор по области — доводом, а не после выдачи: отобранная на клиенте
 * страница выходила бы пустой при том, что подходящие шаблоны есть дальше.
 */
export function listTemplates(
  values: { spaceId?: string; cursor?: string; limit?: number } = {},
  fetcher?: typeof fetch,
  headers?: Record<string, string>
) {
  return post<TemplatePage>('/api/templates/', values, { fetcher, headers });
}

export function templateInfo(
  templateId: string,
  fetcher?: typeof fetch,
  headers?: Record<string, string>
) {
  return post<TemplateBody>('/api/templates/info', { templateId }, { fetcher, headers });
}

/**
 * Завести шаблон.
 *
 * Содержимое берётся из страницы: шаблон и заводят затем, чтобы повторить
 * удачную страницу. Пустой `spaceId` означает шаблон рабочего пространства.
 */
export function createTemplate(
  values: {
    title: string;
    description?: string;
    icon?: string;
    content?: unknown;
    spaceId?: string;
  },
  fetcher?: typeof fetch
) {
  return post<Template>('/api/templates/create', values, { fetcher });
}

/**
 * Изменить шаблон.
 *
 * `move` отдельным признаком, а не пустым `moveToSpaceId`: пустое значение
 * означает «шаблон рабочего пространства», а не «область не меняем», и без
 * признака перенести шаблон в рабочее пространство было бы нечем.
 */
export function updateTemplate(
  values: {
    templateId: string;
    title?: string;
    description?: string;
    icon?: string;
    content?: unknown;
    moveToSpaceId?: string;
    move?: boolean;
  },
  fetcher?: typeof fetch
) {
  return post<Template>('/api/templates/update', values, { fetcher });
}

export function deleteTemplate(templateId: string, fetcher?: typeof fetch) {
  return post<{ success: boolean }>('/api/templates/delete', { templateId }, { fetcher });
}

/**
 * Завести страницу по шаблону.
 *
 * Область обязательна: шаблон рабочего пространства сам по себе не знает, где
 * должна появиться страница.
 */
export function useTemplate(
  values: { templateId: string; spaceId: string; parentPageId?: string },
  fetcher?: typeof fetch
) {
  return post<MadePage>('/api/templates/use', values, { fetcher });
}
